### ISSUE TITLE
A complex field with at least 2 required sub-fields that have conditional visibility, throws error on trying to delete an entry

### ISSUE THREAD

**Issue by ShariqKhan2012** [htmlburger/carbon-fields#738]
## Version

  - Carbon Fields: 3.1.2
  - WordPress: 5.2.2
  - PHP: 7.2.19

## Expected Behavior
User should be able to delete a 'complex field' entry 

## Actual Behavior
If a complex field contains 2 fields that have conditional visibility, and are 'mandatory' fields, then deleting the complex field entry results in an error. All the  fields (including other entries of that complex field) in that container suddenly disappear and actually removed from the DOM.

The browser console also shows this JS error:
`react-dom.min.js?ver=16.8.4:117 TypeError: Cannot read property 'fields' of undefined
    at X (metaboxes.min.js?ver=3.1.2:1)
    at metaboxes.min.js?ver=3.1.2:1
    at core.min.js?ver=3.1.2:10
    at Object.next (vendor.min.js?ver=3.1.2:11)
    at vendor.min.js?ver=3.1.2:11
    at metaboxes.min.js?ver=3.1.2:1
    at vendor.min.js?ver=3.1.2:11
    at core.min.js?ver=3.1.2:10
    at data.min.js?ver=4.4.0:1
    at Array.forEach (<anonymous>)
Bg @ react-dom.min.js?ver=16.8.4:117
c.callback @ react-dom.min.js?ver=16.8.4:126
Ag @ react-dom.min.js?ver=16.8.4:116
zg @ react-dom.min.js?ver=16.8.4:116
qi @ react-dom.min.js?ver=16.8.4:131
ui @ react-dom.min.js?ver=16.8.4:133
(anonymous) @ react-dom.min.js?ver=16.8.4:158
unstable_runWithPriority @ react.min.js?ver=16.8.4:27
Vc @ react-dom.min.js?ver=16.8.4:158
Sc @ react-dom.min.js?ver=16.8.4:158
Z @ react-dom.min.js?ver=16.8.4:156
ah @ react-dom.min.js?ver=16.8.4:159
xf @ react-dom.min.js?ver=16.8.4:40
react-dom.min.js?ver=16.8.4:157 Uncaught TypeError: Cannot read property 'fields' of undefined
    at X (metaboxes.min.js?ver=3.1.2:1)
    at metaboxes.min.js?ver=3.1.2:1
    at core.min.js?ver=3.1.2:10
    at Object.next (vendor.min.js?ver=3.1.2:11)
    at vendor.min.js?ver=3.1.2:11
    at metaboxes.min.js?ver=3.1.2:1
    at vendor.min.js?ver=3.1.2:11
    at core.min.js?ver=3.1.2:10
    at data.min.js?ver=4.4.0:1
    at Array.forEach (<anonymous>)
X @ metaboxes.min.js?ver=3.1.2:1
(anonymous) @ metaboxes.min.js?ver=3.1.2:1
(anonymous) @ core.min.js?ver=3.1.2:10
(anonymous) @ vendor.min.js?ver=3.1.2:11
(anonymous) @ vendor.min.js?ver=3.1.2:11
(anonymous) @ metaboxes.min.js?ver=3.1.2:1
(anonymous) @ vendor.min.js?ver=3.1.2:11
(anonymous) @ core.min.js?ver=3.1.2:10
(anonymous) @ data.min.js?ver=4.4.0:1
n @ data.min.js?ver=4.4.0:1
(anonymous) @ data.min.js?ver=4.4.0:1
v @ data.min.js?ver=4.4.0:1
(anonymous) @ data.min.js?ver=4.4.0:1
(anonymous) @ data.min.js?ver=4.4.0:1
(anonymous) @ data.min.js?ver=4.4.0:1
value @ data.min.js?ver=4.4.0:1
(anonymous) @ core.min.js?ver=3.1.2:10
(anonymous) @ vendor.min.js?ver=3.1.2:11
(anonymous) @ vendor.min.js?ver=3.1.2:11
(anonymous) @ vendor.min.js?ver=3.1.2:11
(anonymous) @ vendor.min.js?ver=3.1.2:1
(anonymous) @ vendor.min.js?ver=3.1.2:1
(anonymous) @ vendor.min.js?ver=3.1.2:1
next @ vendor.min.js?ver=3.1.2:11
(anonymous) @ vendor.min.js?ver=3.1.2:11
(anonymous) @ vendor.min.js?ver=3.1.2:11
r.triggerUnmount @ vendor.min.js?ver=3.1.2:11
o.componentWillUnmount @ vendor.min.js?ver=3.1.2:11
Dg @ react-dom.min.js?ver=16.8.4:119
Eg @ react-dom.min.js?ver=16.8.4:123
ni @ react-dom.min.js?ver=16.8.4:128
ui @ react-dom.min.js?ver=16.8.4:133
(anonymous) @ react-dom.min.js?ver=16.8.4:158
unstable_runWithPriority @ react.min.js?ver=16.8.4:27
Vc @ react-dom.min.js?ver=16.8.4:158
Sc @ react-dom.min.js?ver=16.8.4:158
Z @ react-dom.min.js?ver=16.8.4:156
ah @ react-dom.min.js?ver=16.8.4:159
xf @ react-dom.min.js?ver=16.8.4:40
Show 9 more frames`

## Container definition
```php
$prefix = 'shk_';
	$container_prefix = $prefix . 'testcontainer_';
	$field_prefix = '' ;

	$shk_metabox_key = $container_prefix . 'testcampaign_settings';
	
	$shk_test_container = Container::make( 'post_meta', $shk_metabox_key, 'Test container' );
	$shk_test_container->where( 'post_type', '=', 'page' );
					
	$shk_test_container->add_fields( array(		
		Field::make( 'complex', $field_prefix . 'sports_prefs', 'Sport Preferences' )->set_collapsed(true)
			->add_fields( array(
				
				/* FIRST SET OF FIELDS	*/		
				Field::make( 'checkbox', $field_prefix . 'test_status', 'Active' )
					//->set_option_value('yes')
					->set_classes('shk-checkbox'),
				
				Field::make( 'text', $field_prefix . 'sample_text1', 'Lorem' )
					->set_required( true )
					->set_conditional_logic( array(
						//'relation' => 'AND', 
						array(
							'field' => $field_prefix . 'test_status',
							'value' => true,
							'compare' => '=', 
						)
					) ),

				Field::make( 'text', $field_prefix . 'sample_text2', 'Ipsum' )
					->set_required( true )
					->set_conditional_logic( array(
						//'relation' => 'AND', 
						array(
							'field' => $field_prefix . 'test_status',
							'value' => true,
							'compare' => '=', 
						)
					) ),				

					
		


				/* SECOND SET OF FIELDS	*/		
				/*Field::make( 'select', $field_prefix . 'chosen_sport', 'Pick a sport' )
					->add_options( array(
						'soccer' => 'Soccer',
						'cricket' => 'Cricket',
					) ),					


				Field::make( 'text', $field_prefix . 'soccer_team', 'Favorite soccer team' )
					->set_required( true )
					->set_conditional_logic( array(
						'relation' => 'AND', 
						array(
							'field' => $field_prefix . 'chosen_sport',
							'value' => 'soccer',
							'compare' => '=', 
						)
					) ),

				Field::make( 'text', $field_prefix . 'cricket_team', 'Favorite cricket team' )
					->set_required( true )
					->set_conditional_logic( array(
						'relation' => 'AND', 
						array(
							'field' => $field_prefix . 'chosen_sport',
							'value' => 'cricket',
							'compare' => '=', 
						)
					) )*/
			) )			
	) );
```

The issue is reproducible with the 'FIRST SET OF FIELDS' as well as 'SECOND SET OF FIELDS' 

## Steps to Reproduce the Problem

  1. Add a complex field to a container attached to any post type or page.
  2. Add a checkbox or a dropdown (I haven't tried with other fields) sub-field to the complex field
  3. Add **at least 2** text fields as sub-fields to the complex field. Configure these 2 text-fields to be conditionally visible depending on the value of the checkbox or the dropdown field added in step 2
4. Mark these 2 text fields as **required**.
5. Go to the screen the container is attached to. Eg. the 'Add new post' screen.
6. 'Add entry' for the complex field you added. Now try to delete the entry.

![Issue 1](https://user-images.githubusercontent.com/17094481/61395419-0a939f00-a8e3-11e9-896b-53d9b434c2ae.png)

7. All fields from the container get removed from the DOM and the browser console shows this JS error:
![Issue 2](https://user-images.githubusercontent.com/17094481/61395417-0a939f00-a8e3-11e9-8466-a29b91820710.png)

## Comments
It is absolutely essential that the there should be at least 2 fields that have conditional visibility and are marked as mandatory by set_required(true)
If those fields are not marked as *required* or have normal visibility as opposed to conditional visibility, then the issue does not occur.

**Comment by ShariqKhan2012** [comment]
The error log I got after enabling the SCRIPT_DEBUG flag:

`jquery-migrate.js:23 JQMIGRATE: Migrate is installed with logging active, version 1.4.1
react-dom.js:21347 Download the React DevTools for a better development experience: https://fb.me/react-devtools
react-dom.js:500 Warning: Unsafe lifecycle methods were found within a strict-mode tree:
    in StrictMode (created by Editor)
    in Editor (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithSelect(Editor)

componentWillUpdate: Please update the following components to use componentDidUpdate instead: FillComponent

Learn more about this warning here:
https://fb.me/react-strict-mode-warnings
warningWithoutStack @ react-dom.js:500
webpack://cf.%5Bname%5D/./packages/metaboxes/hocs/with-conditional-logic/index.js?:41 Uncaught TypeError: Cannot read property 'fields' of undefined
    at getFieldsFromFieldsHolder (webpack://cf.%5Bname%5D/./packages/metaboxes/hocs/with-conditional-logic/index.js?:41)
    at output (webpack://cf.%5Bname%5D/./packages/metaboxes/hocs/with-conditional-logic/index.js?:129)
    at eval (webpack://cf.%5Bname%5D/./packages/core/hocs/with-conditional-logic/index.js?:66)
    at Object.eval [as next] (webpack:///./node_modules/refract-callbag/index.es.js?:219)
    at eval (webpack:///./node_modules/callbag-to-obs/readme.js?:53)
    at eval (webpack://cf.%5Bname%5D/./node_modules/callbag-distinct-until-changed/dist/callbag-distinct-until-changed.es.js?:34)
    at eval (webpack:///./node_modules/callbag-merge/readme.js?:48)
    at eval (webpack://cf.%5Bname%5D/./packages/core/utils/from-selector.js?:28)
    at data.js:1362
    at Array.forEach (<anonymous>)
webpack://cf.%5Bname%5D/./packages/metaboxes/hocs/with-conditional-logic/index.js?:41 Uncaught TypeError: Cannot read property 'fields' of undefined
    at getFieldsFromFieldsHolder (webpack://cf.%5Bname%5D/./packages/metaboxes/hocs/with-conditional-logic/index.js?:41)
    at output (webpack://cf.%5Bname%5D/./packages/metaboxes/hocs/with-conditional-logic/index.js?:129)
    at eval (webpack://cf.%5Bname%5D/./packages/core/hocs/with-conditional-logic/index.js?:66)
    at Object.eval [as next] (webpack:///./node_modules/refract-callbag/index.es.js?:219)
    at eval (webpack:///./node_modules/callbag-to-obs/readme.js?:53)
    at eval (webpack://cf.%5Bname%5D/./node_modules/callbag-distinct-until-changed/dist/callbag-distinct-until-changed.es.js?:34)
    at eval (webpack:///./node_modules/callbag-merge/readme.js?:48)
    at eval (webpack://cf.%5Bname%5D/./packages/core/utils/from-selector.js?:28)
    at data.js:1362
    at Array.forEach (<anonymous>)
react-dom.js:17206 The above error occurred in the <WithEffects> component:
    in WithEffects (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(WithEffects) (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(WithDispatch(WithEffects)) (created by WithEffects)
    in WithEffects (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithSelect(WithEffects) (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithConditionalLogic(WithDispatch(WithDispatch(WithEffects))) (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(WithConditionalLogic(WithDispatch(WithDispatch(WithEffects)))) (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithField(WithConditionalLogic(WithDispatch(WithDispatch(WithEffects)))) (created by WithFilters(WithFilters(TextField)))
    in WithFilters(WithFilters(TextField)) (created by ComplexGroup)
    in div (created by Field)
    in div (created by Field)
    in Field
    in Unknown (created by WithFilters(Field))
    in WithFilters(Field) (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithSelect(WithFilters(Field)) (created by WithFilters(WithSelect(WithFilters(Field))))
    in WithFilters(WithSelect(WithFilters(Field))) (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(WithFilters(WithSelect(WithFilters(Field)))) (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithField(WithFilters(WithSelect(WithFilters(Field)))) (created by ComplexGroup)
    in div (created by ComplexGroup)
    in div (created by ComplexGroup)
    in ComplexGroup (created by ComplexField)
    in div (created by ComplexField)
    in Sortable (created by ComplexField)
    in ComplexField (created by ComplexField)
    in ComplexField
    in Unknown (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(Component) (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithSelect(WithDispatch(Component)) (created by WithFilters(ComplexField))
    in WithFilters(ComplexField) (created by WithEffects)
    in WithEffects (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(WithEffects) (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(WithDispatch(WithEffects)) (created by WithEffects)
    in WithEffects (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithSelect(WithEffects) (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithConditionalLogic(WithDispatch(WithDispatch(WithEffects))) (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(WithConditionalLogic(WithDispatch(WithDispatch(WithEffects)))) (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithField(WithConditionalLogic(WithDispatch(WithDispatch(WithEffects)))) (created by WithFilters(WithFilters(ComplexField)))
    in WithFilters(WithFilters(ComplexField)) (created by Container)
    in div (created by Field)
    in div (created by Field)
    in Field
    in Unknown (created by WithFilters(Field))
    in WithFilters(Field) (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithSelect(WithFilters(Field)) (created by WithFilters(WithSelect(WithFilters(Field))))
    in WithFilters(WithSelect(WithFilters(Field))) (created by ComponentWithDispatch)
    in ComponentWithDispatch (created by Context.Consumer)
    in WithDispatch(WithFilters(WithSelect(WithFilters(Field)))) (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithField(WithFilters(WithSelect(WithFilters(Field)))) (created by Container)
    in div (created by Container)
    in div (created by Container)
    in Container (created by WithFilters(Container))
    in WithFilters(Container) (created by ComponentWithSelect)
    in ComponentWithSelect (created by Context.Consumer)
    in WithContainer(WithFilters(Container))

Consider adding an error boundary to your tree to customize error handling behavior.
Visit https://fb.me/react-error-boundaries to learn more about error boundaries.
logCapturedError @ react-dom.js:17206
react-dom.js:20401 Uncaught TypeError: Cannot read property 'fields' of undefined
    at getFieldsFromFieldsHolder (webpack://cf.%5Bname%5D/./packages/metaboxes/hocs/with-conditional-logic/index.js?:41)
    at output (webpack://cf.%5Bname%5D/./packages/metaboxes/hocs/with-conditional-logic/index.js?:129)
    at eval (webpack://cf.%5Bname%5D/./packages/core/hocs/with-conditional-logic/index.js?:66)
    at Object.eval [as next] (webpack:///./node_modules/refract-callbag/index.es.js?:219)
    at eval (webpack:///./node_modules/callbag-to-obs/readme.js?:53)
    at eval (webpack://cf.%5Bname%5D/./node_modules/callbag-distinct-until-changed/dist/callbag-distinct-until-changed.es.js?:34)
    at eval (webpack:///./node_modules/callbag-merge/readme.js?:48)
    at eval (webpack://cf.%5Bname%5D/./packages/core/utils/from-selector.js?:28)
    at data.js:1362
    at Array.forEach (<anonymous>)`

**Comment by ShariqKhan2012** [comment]
Just wanted to check if there is a workaround for it if we don't have a fix right now

**Comment by vvasilev-** [comment]
Fixed in v3.1.6.
